patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java

```diff
@@ -103,16 +128,44 @@ public void addRecords(List<CheckstyleRecord> newRecords,
      */
     private static List<CheckstyleRecord> produceDiff(
             List<CheckstyleRecord> list1, List<CheckstyleRecord> list2) {
+        final List<CheckstyleRecord> diff;
+        try {
+            diff = produceDiffEx(list1, list2);
+            diff.addAll(produceDiffEx(list2, list1));
+        }
+        catch (InterruptedException | ExecutionException ex) {
+            throw new IllegalStateException("Multi-threading failure reported", ex);
+        }
+
+        return diff;
+    }
+
+    private static List<CheckstyleRecord> produceDiffEx(
+            List<CheckstyleRecord> list1, List<CheckstyleRecord> list2)
+            throws InterruptedException, ExecutionException {
         final List<CheckstyleRecord> diff = new ArrayList<>();
-        for (CheckstyleRecord rec1 : list1) {
-            if (!isInList(list2, rec1)) {
-                diff.add(rec1);
+        if (list1.size() < SPLIT_SIZE) {
+            for (CheckstyleRecord rec1 : list1) {
+                if (!isInList(list2, rec1)) {
+                    diff.add(rec1);
+                }
             }
         }
-        for (CheckstyleRecord rec2 : list2) {
```

Comment:
Why this loop is removed?

Reply:
This is the loop that amounts to O(N^2) complexity and is the real slowdown that needs multi-threading. The nested loop is inside `isInList`. It is now transferred to line 159 to split the list into pieces and send to the threads to complete. The rest of the loop is moved to line 270 which is inside the thread.

Reply:
I don't get it at all. The code on line 159 will be executed only if the size of the list >= `SPLIT_SIZE`. Otherwise, the method will return only records that are in the list1 and are missing in the list2.

Reply:
It sounds like you are missing https://github.com/checkstyle/contribution/blob/44f02903040bc7f16c834d26ed324e96b2c463a9/patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java#L133-L134 which calls the same method with the lists swapped.